File: artifacts/api-server/src/lib/fabricshift/mapping.ts
import { randomUUID } from "crypto";
import { WORKSPACES, PIPELINES, SQL_OBJECTS } from "./inventory.js";

interface MappingRequest {
  domain_filter?: string | null;
  asset_ids?: string[] | null;
}

export interface MappingResult {
  mapping_id: string;
  source_asset_id: string;
  source_asset_name: string;
  source_asset_type: string;
  business_domain: string;
  recommended_fabric_target: string;
  target_zone: string;
  migration_effort: string;
  risk_level: string;
  blockers: string[];
  recommended_action: string;
  rationale: string;
  medallion_layer: string;
}

function domainFromSchema(schema: string): string {
  const map: Record<string, string> = {
    reg: "Regulatory Reporting",
    claims: "Claims Finance",
    retail: "Retail Sales",
    customer: "Customer Analytics",
    ops: "Operations Monitoring",
  };
  return map[schema] ?? "Unknown";
}

function mapPipeline(p: (typeof PIPELINES)[0]): MappingResult {
  const isHighComplexity = p.complexity === "high";
  const hasManyDeps = p.dependency_count >= 4;
  const isEvent = p.trigger_type === "event";
  const isFailed = p.last_run_status === "failed";

  let risk_level = "low";
  let migration_effort = "low";
  let fabric_target = "Fabric Data Pipeline";
  const target_zone = "Bronze";
  const medallion_layer = "bronze";
  const blockers: string[] = [];
  let recommended_action =
    "Assess pipeline activities and map to Fabric Data Pipeline. Validate trigger compatibility.";
  let rationale =
    "Straightforward ingestion pipeline with low complexity. Map to Fabric Data Pipeline with minimal rework.";

  if (isHighComplexity) {
    migration_effort = "high";
    risk_level = "medium";
    fabric_target = "Fabric Data Pipeline + Dataflow Gen2";
    rationale =
      "High activity count requires decomposition into Fabric Data Pipeline orchestration with Dataflow Gen2 for transformation steps.";
  }

  if (hasManyDeps) {
    risk_level = isHighComplexity ? "high" : "medium";
    blockers.push("High dependency count — dependency ordering must be validated before migration.");
    rationale += " Upstream dependencies must be migrated first.";
  }

  if (isEvent) {
    blockers.push(
      "Event-based trigger — Fabric event-driven pipeline support must be verified for this trigger source.",
    );
    risk_level = "medium";
  }

  if (isFailed) {
    blockers.push(
      "Last run failed — pipeline stability must be resolved before migration to avoid propagating failures.",
    );
    risk_level = risk_level === "high" ? "critical" : "high";
  }

  if (p.average_runtime_minutes > 60) {
    migration_effort = "high";
    rationale +=
      " Long average runtime suggests complex transformations — consider Spark Notebook for heavy compute steps.";
  }

  return {
    mapping_id: `map-${p.pipeline_id}`,
    source_asset_id: p.pipeline_id,
    source_asset_name: p.pipeline_name,
    source_asset_type: "ADF Pipeline",
    business_domain: p.business_domain,
    recommended_fabric_target: fabric_target,
    target_zone,
    migration_effort,
    risk_level,
    blockers,
    recommended_action,
    rationale,
    medallion_layer,
  };
}

function mapSqlObject(o: (typeof SQL_OBJECTS)[0]): MappingResult {
  const isLarge = o.size_mb > 500;
  const hasSensitive = o.sensitive_data_flag;
  const isView = o.object_type === "view";
  const hasNoOwner = !o.owner;
  const hasDownstreamReports = o.downstream_reports > 3;

  let risk_level = "low";
  let migration_effort = "low";
  let fabric_target = "Fabric Lakehouse Delta Table";
  let target_zone = "Silver";
  let medallion_layer = "silver";
  const blockers: string[] = [];
  let recommended_action = "";
  let rationale = "";

  if (isView) {
    fabric_target = "Fabric Warehouse View";
    target_zone = "Gold";
    medallion_layer = "gold";
    recommended_action =
      "Re-implement as a semantic layer view or reporting mart in Fabric Warehouse Gold zone.";
    rationale =
      "Views typically represent aggregated or conformed entities. Map to Gold layer in Fabric Warehouse.";
  } else if (o.object_name.startsWith("stg_") || o.object_name.startsWith("raw_")) {
    target_zone = "Bronze";
    medallion_layer = "bronze";
    fabric_target = "Fabric Lakehouse Delta Table (Bronze)";
    recommended_action = "Land raw data to OneLake Bronze zone as Delta table. Preserve source schema.";
    rationale = "Staging/raw tables map to Bronze landing zone in the medallion architecture.";
  } else {
    recommended_action =
      "Migrate table to Fabric Lakehouse Silver zone as Delta table with partition strategy.";
    rationale =
      "Conformed fact/dimension tables map to Silver zone. Apply column-level data typing and null constraints.";
  }

  if (isLarge) {
    migration_effort = "high";
    risk_level = "medium";
    blockers.push(
      `Large table (${Math.round(o.size_mb)} MB) — validate OneLake partition strategy before migration.`,
    );
  }

  if (hasSensitive) {
    risk_level = risk_level === "low" ? "medium" : risk_level;
    blockers.push(
      "Sensitive data flag set — data classification and Fabric information protection labels required.",
    );
  }

  if (hasNoOwner) {
    blockers.push("No owner assigned — ownership must be established before migration.");
    risk_level = "high";
  }

  if (hasDownstreamReports) {
    blockers.push(
      "High downstream report dependency — report migration must be coordinated to prevent analytics outage.",
    );
    migration_effort = migration_effort === "low" ? "medium" : "high";
  }

  if (!o.primary_key && !isView) {
    blockers.push("No primary key — Delta table identity column strategy must be defined.");
    risk_level = risk_level === "low" ? "medium" : risk_level;
  }

  return {
    mapping_id: `map-${o.object_id}`,
    source_asset_id: o.object_id,
    source_asset_name: `${o.schema_name}.${o.object_name}`,
    source_asset_type: o.object_type === "view" ? "SQL View" : "SQL Table",
    business_domain: domainFromSchema(o.schema_name),
    recommended_fabric_target: fabric_target,
    target_zone,
    migration_effort,
    risk_level,
    blockers,
    recommended_action,
    rationale,
    medallion_layer,
  };
}

let cachedResults: MappingResult[] = [];

export function computeMappingRun(request?: MappingRequest) {
  const results: MappingResult[] = [];

  const pipelines = request?.domain_filter
    ? PIPELINES.filter((p) => p.business_domain === request.domain_filter)
    : PIPELINES;

  const sqlObjects = request?.domain_filter
    ? SQL_OBJECTS.filter((o) => domainFromSchema(o.schema_name) === request.domain_filter)
    : SQL_OBJECTS;

  for (const p of pipelines) {
    if (!request?.asset_ids || request.asset_ids.includes(p.pipeline_id)) {
      results.push(mapPipeline(p));
    }
  }

  for (const o of sqlObjects) {
    if (!request?.asset_ids || request.asset_ids.includes(o.object_id)) {
      results.push(mapSqlObject(o));
    }
  }

  const effort: Record<string, number> = { low: 0, medium: 0, high: 0 };
  const risk: Record<string, number> = { low: 0, medium: 0, high: 0, critical: 0 };
  for (const r of results) {
    effort[r.migration_effort] = (effort[r.migration_effort] ?? 0) + 1;
    risk[r.risk_level] = (risk[r.risk_level] ?? 0) + 1;
  }

  const run = {
    run_id: randomUUID(),
    run_at: new Date().toISOString(),
    total_mapped: results.length,
    results,
    effort_distribution: effort,
    risk_distribution: risk,
  };

  cachedResults = results;
  return run;
}

export function getCachedResults(): MappingResult[] {
  if (cachedResults.length === 0) {
    computeMappingRun();
  }
  return cachedResults;
}

export function getCachedResultById(mappingId: string): MappingResult | undefined {
  if (cachedResults.length === 0) {
    computeMappingRun();
  }
  return cachedResults.find((r) => r.mapping_id === mappingId);
}

File: fabricshift/fabricshift/mapping.py
"""Fabric target mapping engine — maps Synapse assets to Fabric targets."""
from __future__ import annotations
import uuid
from .models import (
    SynapseWorkspace, AdfPipeline, SqlObject, PowerBiReport,
    DataProduct, MappingResult
)


def _effort_from_complexity(complexity: str) -> str:
    return {"high": "high", "medium": "medium", "low": "low"}.get(complexity, "medium")


def map_workspace(w: SynapseWorkspace) -> MappingResult:
    target = "Fabric Workspace (Lakehouse + Warehouse)"
    zone = "Analytics"
    effort = "high" if w.sql_pool_count > 1 else "medium"
    risk = "high" if w.criticality in ("critical", "high") else "medium"
    blockers: list[str] = []
    if not w.migration_candidate:
        blockers.append("Workspace not flagged as a migration candidate")
    return MappingResult(
        mapping_id=str(uuid.uuid4()),
        source_asset_name=w.workspace_name,
        source_asset_type="Synapse Workspace",
        business_domain=w.business_domain,
        recommended_fabric_target=target,
        target_zone=zone,
        migration_effort=effort,
        risk_level=risk,
        blockers=blockers,
        recommended_action="Provision Fabric workspace, enable OneLake, migrate SQL pools",
        rationale="Synapse dedicated SQL pools map directly to Fabric Warehouse; Spark pools map to Fabric notebooks/lakehouses",
        medallion_layer="source",
    )


def map_pipeline(p: AdfPipeline) -> MappingResult:
    effort = _effort_from_complexity(p.complexity)
    risk = "critical" if p.dependency_count > 5 else "high" if p.dependency_count > 2 else "medium"
    blockers: list[str] = []
    if p.last_run_status == "failed":
        blockers.append("Pipeline last run failed — must resolve before migration")
    return MappingResult(
        mapping_id=str(uuid.uuid4()),
        source_asset_name=p.pipeline_name,
        source_asset_type="ADF Pipeline",
        business_domain=p.business_domain,
        recommended_fabric_target="Fabric Data Pipeline (ADF-compatible)",
        target_zone="Ingestion",
        migration_effort=effort,
        risk_level=risk,
        blockers=blockers,
        recommended_action="Re-create pipeline in Fabric using the ADF-compatible pipeline editor",
        rationale=f"ADF-compatible pipelines supported natively in Fabric. {p.activity_count} activities, {p.dependency_count} dependencies.",
        medallion_layer="pipeline",
    )


def map_sql_object(o: SqlObject) -> MappingResult:
    effort = "high" if o.size_mb > 500 else "medium"
    risk = "critical" if o.sensitive_data_flag else "high" if o.downstream_reports > 3 else "medium"
    blockers: list[str] = []
    if o.sensitive_data_flag:
        blockers.append("Contains sensitive/PII data — Purview classification required before migration")
    layer = "gold" if o.object_type in ("VIEW", "MATERIALIZED_VIEW") else "silver"
    return MappingResult(
        mapping_id=str(uuid.uuid4()),
        source_asset_name=f"{o.schema_name}.{o.object_name}",
        source_asset_type=f"SQL {o.object_type}",
        business_domain="Unknown",
        recommended_fabric_target="Fabric Warehouse (Delta table)" if o.object_type == "TABLE" else "Fabric Lakehouse (Delta view)",
        target_zone="Warehouse",
        migration_effort=effort,
        risk_level=risk,
        blockers=blockers,
        recommended_action="Use CTAS or CETAS to migrate to Fabric Warehouse / Lakehouse Delta tables",
        rationale=f"{o.row_count:,} rows, {o.size_mb} MB. {o.downstream_reports} downstream reports.",
        medallion_layer=layer,
    )


def map_report(r: PowerBiReport) -> MappingResult:
    risk = "critical" if r.row_level_security and r.direct_query_flag else "high" if r.direct_query_flag else "medium"
    blockers: list[str] = []
    if r.direct_query_flag:
        blockers.append("DirectQuery mode — must migrate dataset to Fabric before report migration")
    if r.row_level_security:
        blockers.append("Row-level security rules must be re-validated in Fabric")
    return MappingResult(
        mapping_id=str(uuid.uuid4()),
        source_asset_name=r.report_name,
        source_asset_type="Power BI Report",
        business_domain="Unknown",
        recommended_fabric_target="Fabric Power BI Report (DirectLake)",
        target_zone="Reporting",
        migration_effort="medium",
        risk_level=risk,
        blockers=blockers,
        recommended_action="Migrate dataset to Fabric Lakehouse, switch report to DirectLake mode",
        rationale=f"Dataset: {r.dataset_name}. Refresh: {r.refresh_frequency}. DirectQuery: {r.direct_query_flag}.",
        medallion_layer="report",
    )
